### ISSUE TITLE
Code for JavaScript steps is always stored twice during exports

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#2393]
Since the data object should always be equal to the script object in a JavaScript Step, there should be no need to store it twice. It's usually not a problem, but it becomes cumbersome for channels that have several dozen steps (or even hundreds).

Imported Issue. Original Details:
Jira Issue Key: MIRTH-2457
Reporter: narupley
Created: 2013-04-02T09:24:37.000-0700

**Comment by rbeckman-nextgen** [comment]
We have channels with over 50 transformers.  The XML export is extremely valuable for a read-only review of the "life cycle" of variables, function calls, etc.  (Rather than having to open each transformer one at a time in Mirth Connect for such a full code review). 
It would be hugely beneficial if the JavaScript code was stored only once in the XML export.  Most of our code is Javascript and it is very difficult to parse a Mirth XML export file that has over a thousand lines of extra duplicate code.  I vote a big YES for this change.

Imported Comment. Original Details:
Author: moehinder
Created: 2013-04-04T06:39:18.000-0700

**Comment by rbeckman-nextgen** [comment]
In order to do this, we'll have to convert the getScript method in Step to call out to the plugin type and generate the script from data. We'll also have to move the script generation code from Client to Shared by creating a ScriptGenerator class for each step/rule type. This needs to get called on both the client (viewing the code tab) and the server (builder the channel scripts). This also means that non of the step/rules apart from JavaScript will actually have their script saved in the channel export. 

However I suspect that some people may want to see all the code stored in the channel export based off of John's comment. However, MIRTH-2128 may be able to resolve that since users will be able to view their code for all step/rule types directly in Mirth Connect.

Another issue is that if we no longer store the script along with each channel, then if the channel generation code changes for a step/rule type with a new version of MC, then the newly generated code will be used upon upgrading even though the user never actually saved the channel again. This might actually be a plus because if the generation code ever changed, the next time the user saved the channel that code would get updated anyway. So this way it would be consistent with the version immediately upon migration. Since it won't affect the actual JavaScript step/rule, I think that this is ok.

Imported Comment. Original Details:
Author: wayneh
Created: 2014-12-22T19:29:53.000-0800

**Comment by rbeckman-nextgen** [comment]
Anyone have additional thoughts on this? Basically we can't have it both ways. Either we store the code twice for JavaScript steps or the code for all step types gets generated on deploy and may change (however unlikely) after an upgrade.

Imported Comment. Original Details:
Author: wayneh
Created: 2015-01-27T12:42:08.000-0800

**Comment by rbeckman-nextgen** [comment]
I think it's more correct to generate the script on deploy.

Imported Comment. Original Details:
Author: narupley
Created: 2015-01-27T13:02:27.000-0800

**Comment by rbeckman-nextgen** [comment]
This is already fixed with the changes in MIRTH-3025.

Imported Comment. Original Details:
Author: narupley
Created: 2017-01-29T10:29:40.000-0800
